auditing/meilisearch.go

```diff
@@ -0,0 +1,221 @@
+	client           *meilisearch.Client
+	index            *meilisearch.Index
+	log              *zap.SugaredLogger
+	indexPrefix      string
+	rotationInterval Interval
+	keep             int64
+}
+
+func New(c Config) (Auditing, error) {
+	client := meilisearch.NewClient(meilisearch.ClientConfig{
+		Host:   c.URL,
+		APIKey: c.APIKey,
+	})
+	v, err := client.GetVersion()
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to meilisearch at:%s %w", c.URL, err)
+	}
+	c.Log.Infow("meilisearch", "connected to", v, "index rotated", c.RotationInterval, "index keep", c.Keep)
+
+	index := client.Index(c.IndexPrefix)
+	if c.RotationInterval != "" {
+		index = client.Index(indexName(c.IndexPrefix, c.RotationInterval))
+	}
+	_, err = index.UpdateTypoTolerance(&meilisearch.TypoTolerance{
+		Enabled: false,
+	})
+	if err != nil {
+		return nil, err
+	}
+	_, err = index.UpdateSortableAttributes(&[]string{"timestamp-unix", "timestamp"})
+	if err != nil {
+		return nil, err
+	}
+
+	a := &meiliAuditing{
+		component:        c.Component,
+		client:           client,
+		index:            index,
+		log:              c.Log.Named("auditing"),
+		indexPrefix:      c.IndexPrefix,
+		rotationInterval: c.RotationInterval,
+		keep:             c.Keep,
+	}
+	if a.component == "" {
+		ex, err := os.Executable()
+		if err != nil {
+			return nil, err
+		}
+		a.component = filepath.Base(ex)
+	}
+
+	if c.RotationInterval != "" {
+		// create a new Index every interval
+		cn := cron.New()
+		err := cn.AddFunc(string(c.RotationInterval), a.newIndex)
+		if err != nil {
+			return nil, err
+		}
+		cn.Start()
+	}
+	return a, nil
+}
+
+func (a *meiliAuditing) Index(entry Entry) error {
+	if entry.Id == "" {
+		entry.Id = uuid.NewString()
+	}
+	if entry.Component == "" {
+		entry.Component = a.component
+	}
+	if entry.Timestamp.IsZero() {
+		entry.Timestamp = time.Now()
+	}
+	doc := make(map[string]any)
+	doc["id"] = entry.Id
+	doc["component"] = entry.Component
+	if entry.Type != "" {
+		doc["type"] = string(entry.Type)
+	}
+	doc["timestamp"] = entry.Timestamp.Format(time.RFC3339)
+	doc["timestamp-unix"] = entry.Timestamp.Unix()
+	if entry.User != "" {
+		doc["user"] = entry.User
+	}
+	if entry.Tenant != "" {
+		doc["tenant"] = entry.Tenant
+	}
+	if entry.RequestId != "" {
+		doc["rqid"] = entry.RequestId
+	}
+	if entry.Detail != "" {
+		doc["detail"] = string(entry.Detail)
+	}
+	if entry.Phase != "" {
+		doc["phase"] = string(entry.Phase)
+	}
+	if entry.Path != "" {
+		doc["path"] = entry.Path
+	}
+	if entry.ForwardedFor != "" {
+		doc["forwarded-for"] = entry.ForwardedFor
+	}
+	if entry.RemoteAddr != "" {
+		doc["remote-addr"] = entry.RemoteAddr
+	}
+	if entry.StatusCode != 0 {
+		doc["status-code"] = entry.StatusCode
+	}
+	if entry.Error != nil {
+		doc["error"] = entry.Error.Error()
+	}
+	if entry.Body != nil {
+		doc["body"] = entry.Body
+	}
+
+	for k, v := range doc {
+		a.log.Debugw("index", "key", k, "value", v, "type", fmt.Sprintf("%T", v))
+	}
```

Comment:
Adding a loop just for debugging maybe a bit much. I think `Debugw("debug", "doc", doc)` has the same effect. Or check if debug is enabled with `!a.log.Level().Enabled(zapcore.DebugLevel)` and then marshal a JSON to print in order to be able to see the serialized content.

Reply:
Good catch! This was a debugging leftover.